cpu/esp8266/periph/rtt.c

```diff
@@ -0,0 +1,201 @@
+
+extern uint32_t pm_rtc_clock_cali_proc(void);
+extern uint32_t pm_rtc2usec(uint32_t rtc_cycles, uint32_t period);
+
+void IRAM rtt_cb(void *arg)
+{
+    if (rtt_config.alarm && rtt_config.alarm_cb) {
+        /* reset the alarm first */
+        rtt_cb_t alarm_cb = rtt_config.alarm_cb;
+        void * alarm_arg = rtt_config.alarm_arg;
+        rtt_clear_alarm();
+        frc2.alarm = 0;
+        /* call the alarm handler afterwards */
+        alarm_cb(alarm_arg);
+    }
+    else if (rtt_config.overflow_cb) {
+        rtt_config.overflow_cb(rtt_config.overflow_arg);
+    }
+}
+
+void rtt_init(void)
+{
+    DEBUG("%s saved rtt=%u rtc=%u\n",
+          __func__, _rtt_counter_saved, _rtc_counter_saved);
+
+    frc2.ctrl.clk_div = FRC2_CLK_DIV_256;
+    frc2.ctrl.reload = 0;
+    frc2.ctrl.intr_hold = 0;
+    frc2.ctrl.enable = 1;
+
+    if (_rtt_counter_saved || _rtc_counter_saved) {
+        /* if in init after power on, restore the RTT counter value */
+        uint32_t rtc_diff = RTC.COUNTER - _rtc_counter_saved;
+        uint32_t rtc_diff_us = pm_rtc2usec(rtc_diff, pm_rtc_clock_cali_proc());
+        uint32_t rtt_diff = RTT_US_TO_TICKS(rtc_diff_us);
+
+        frc2.load = _rtt_counter_saved + rtt_diff;
+
+        DEBUG("%s rtc_diff=%u rtt_diff=%u load=%u\n", __func__,
+              rtc_diff, rtt_diff, _rtt_counter_saved + rtt_diff);
+    }
+    else {
+        frc2.load = 0;
+        DEBUG("%s after power on\n", __func__);
+    }
+
+    /* emulate overflow interrupt */
+    frc2.alarm = 0;
+
+    ets_isr_attach (ETS_FRC2_INUM, rtt_cb, NULL);
+    ets_isr_unmask (BIT(ETS_FRC2_INUM));
+    DPORT.INT_ENABLE |= DPORT_INT_ENABLE_FRC2;
+}
+
+void rtt_poweron(void)
+{
+    /* power on simply reactivates the FRC2 counter */
+    frc2.ctrl.enable = 1;
+}
+
+void rtt_poweroff(void)
+{
+    /* power off simply deactivates the FRC2 counter */
+    frc2.ctrl.enable = 0;
+}
+
+void rtt_set_overflow_cb(rtt_cb_t cb, void *arg)
+{
+    /* there is no overflow interrupt, we emulate */
+    rtt_config.overflow_cb = cb;
+    rtt_config.overflow_arg = arg;
+}
+
+void rtt_clear_overflow_cb(void)
+{
+    /* there is no overflow interrupt, we emulate */
+    rtt_config.overflow_cb = NULL;
+    rtt_config.overflow_arg = NULL;
+}
+
+uint32_t rtt_get_counter(void)
+{
+    return frc2.count;
+}
+
+void rtt_set_counter(uint32_t counter)
+{
+    frc2.load = counter;
+}
+
+void rtt_set_alarm(uint32_t alarm, rtt_cb_t cb, void *arg)
+{
+    rtt_config.alarm = alarm;
+    rtt_config.alarm_cb = cb;
+    rtt_config.alarm_arg = arg;
+
+    frc2.alarm = alarm;
+}
+
+uint32_t rtt_get_alarm(void)
+{
+    return rtt_config.alarm;
+}
+
+void rtt_clear_alarm(void)
+{
+    frc2.alarm = 0;
+    rtt_config.alarm = 0;
+    rtt_config.alarm_cb = NULL;
+    rtt_config.alarm_arg = NULL;
+}
+
+void rtt_save_counter(void)
+{
+    /* save counters before going to sleep or reboot */
+    _rtt_counter_saved = frc2.count;
+    _rtc_counter_saved = RTC.COUNTER;
```

Comment:
Where does `RTC.COUNTER` come from?

Reply:
In contrast to all other peripherals for which the SDK defines the [registers as a struct](https://github.com/espressif/ESP8266_RTOS_SDK/tree/release/v3.1/components/esp8266/include/esp8266), the registers of the RTC module aren't defined by the SDK at all, see [rtc_register.h](https://github.com/espressif/ESP8266_RTOS_SDK/blob/release/v3.1/components/esp8266/include/esp8266/rtc_register.h).

Therefore, another [piece of vendor code](https://github.com/RIOT-OS/RIOT/blob/master/cpu/esp8266/vendor/esp/rtc_regs.h) is used which was initially extracted from the [esp-open-rtos](https://github.com/SuperHouse/esp-open-rtos). It tries to define the RTC registers as a struct. BTW, this struct is defined exactly in the way you mentioned with volatile members and without linker stuff.

ESP8266 is a terrible platform. It has the worst documentation I have ever seen and there is no consistent development environment.